addons/product_expiry/models/production_lot.py

```diff
@@ -55,3 +55,56 @@ def _onchange_product(self):
         dates_dict = self._get_dates()
         for field, value in dates_dict.items():
             setattr(self, field, value)
+
+
+    def _alert_date_exceeded(self):
+        # Run by a cron
+        # if the alert_date is in the past, log a next activity on the next.production.lot 
+        alert_lot_ids = self.env['stock.production.lot'].search([('alert_date', '<=', fields.Date.today())] )
+        mail_activity_type =self.env.ref('product_expiry.mail_activity_alert_date_reached').id
+        stock_quants = self.env['stock.quant'].search([
+            ('lot_id', 'in', alert_lot_ids.ids),
+            ('quantity', '>', 0)]).filtered(lambda quant: quant.location_id.usage == 'internal' )
+
+        # only for the products that do not have already an activity
+        stock_quants = stock_quants.filtered(lambda quant: 
+            self.env['mail.activity'].search_count([
+                ('res_model', '=', 'stock.production.lot'),
+                ('res_id', '=', quant.lot_id.id),
+                ('activity_type_id','=',mail_activity_type)]) == 0 )
+
+        # when an activity is deleted, it goes into message
+        stock_quants = stock_quants.filtered(lambda quant: 
+            self.env['mail.message'].search_count([
+                ('model', '=', 'stock.production.lot'),
+                ('res_id', '=', quant.lot_id.id),
+                ('subtype_id', '=', 3),
+                ('mail_activity_type_id','=',mail_activity_type)]) == 0 )
+        stock_quants.mapped('lot_id')._log_next_activity_alert_date_reached( 'The alert date has been reached for this lot/serial number' )
+            
+    def _log_next_activity_alert_date_reached(self, note):
```

Comment:
Use the method ``activity_schedule`` defined in mail/models/activity.py instead. Some support already exist for this kind of actions.